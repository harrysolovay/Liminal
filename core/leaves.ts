import { Ty } from "./Ty.ts"

export type none<P extends string = string> = Ty<null, P>
export const none: Ty<null, never> = Ty((description) => ({
  type: "null",
  description,
}))

export type bool<P extends string = string> = Ty<boolean, P>
export const bool: Ty<boolean, never> = Ty((description) => ({
  type: "boolean",
  description,
}))

export type num<P extends string = string> = Ty<number, P>
export const num: Ty<number, never> = Ty((description) => ({
  type: "number",
  description,
}))

export type str<P extends string = string> = Ty<string, P>
export const str: Ty<string, never> = Ty((description) => ({
  type: "string",
  description,
}))
