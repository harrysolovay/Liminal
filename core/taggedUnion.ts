import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import type { Expand } from "../util/type_util.ts"

// TODO: allow numbers as tags + tag keys?
export function taggedUnion<
  K extends string,
  M extends Expand<Record<string, Type | null>>,
>(
  tagKey: K,
  members: M,
): Type<
  { [V in keyof M]: ({ [_ in K]: V } & (M[V] extends Type ? { value: M[V]["T"] } : {})) }[keyof M],
  {},
  Extract<M[keyof M], Type>["P"]
> {
  const entries = Object.entries(members)
  return declare({
    name: "taggedUnion",
    source: {
      factory: taggedUnion,
      args: { tagKey, members },
    },
    subschema: (visit) => ({
      discriminator: tagKey,
      anyOf: entries.map(([k, v]) => ({
        type: "object",
        properties: {
          [tagKey]: {
            type: "string",
            const: k,
          },
          ...(v === null ? {} : { value: visit(v) }),
        },
        required: [tagKey, ...v === null ? [] : ["value"]],
        additionalProperties: false,
      })),
    }),
    output: (f) =>
      f<{ [V in keyof M]: ({ [_ in K]: V } & { value?: unknown }) }[keyof M]>({
        visitor: (value, visit, path) => {
          const tag = value[tagKey]
          const type = members[tag]!
          return ({
            [tagKey]: value[tagKey],
            // TODO: fix typing of `tag`.
            ..."value" in value
              ? {
                value: visit(
                  value.value,
                  type,
                  path.value("value").type(tag as string),
                ),
              }
              : {},
          }) as never
        },
      }),
  })
}
