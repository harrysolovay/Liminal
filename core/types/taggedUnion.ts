import type { Expand } from "../../util/mod.ts"
import { type AnyType, Type } from "../Type.ts"

export function taggedUnion<
  K extends number | string,
  M extends Record<number | string, AnyType | undefined>,
>(
  tagKey: K,
  members: M,
): Type<
  {
    [V in keyof M]: Expand<
      (
        & { [_ in K]: V }
        & (M[V] extends AnyType ? { value: M[V]["T"] } : {})
      )
    >
  }[keyof M],
  Extract<M[keyof M], AnyType>["P"]
> {
  return Type({
    kind: "taggedUnion",
    factory: taggedUnion,
    args: [tagKey, members],
    argsLookup: { tagKey, members },
  })
}
