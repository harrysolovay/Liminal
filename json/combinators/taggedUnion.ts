import { type AnyType, Type } from "../../core/mod.ts"
import type { Expand } from "../../util/type_util.ts"

export function taggedUnion<
  K extends number | string,
  M extends Record<number | string, AnyType | null>,
>(
  tagKey: K,
  members: M,
): Type<
  {
    [V in keyof M]: Expand<({ [_ in K]: V } & (M[V] extends AnyType ? { value: M[V]["T"] } : {}))>
  }[keyof M],
  Extract<M[keyof M], AnyType>["P"]
> {
  return Type({
    factory: taggedUnion,
    args: [tagKey, members],
  })
}
