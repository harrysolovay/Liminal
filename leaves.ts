import { Ty } from "./Ty.ts"

export const none = Ty<null>((description) => ({
  type: "null",
  description,
}))

export const boolean = Ty<boolean>((description) => ({
  type: "boolean",
  description,
}))

export const number = Ty<number>((description) => ({
  type: "number",
  description,
}))

export const string = Ty<string>((description) => ({
  type: "string",
  description,
}))
