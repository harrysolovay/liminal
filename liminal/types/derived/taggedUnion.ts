import type { JSONObjectType } from "../_object.ts"
import type { JSONConstType } from "../const.ts"
import type { JSONStringType } from "../string.ts"
import { type Type } from "../Type.ts"
import { type JSONUnionType, union } from "../union.ts"
import { type NormalizeObjectValueJ, type NormalizeObjectValueT, object, type ObjectFieldValue } from "./object.ts"

export function taggedUnion<const M extends TaggedUnionMembers, K extends Extract<keyof M, string>>(members: M): Type<
  {
    [L in K]: {
      key: L
      value: NormalizeObjectValueT<M[L]>
    }
  }[K],
  JSONUnionType<
    {
      [L in K]: JSONObjectType<{
        key: JSONConstType<JSONStringType, L>
        value: NormalizeObjectValueJ<M[L]>
      }>
    }[K]
  >
> {
  return union(
    ...Object.entries(members).map(([key, value]) => object({ key, value })),
  ) as never
}

export type TaggedUnionMembers = Record<string, ObjectFieldValue>
