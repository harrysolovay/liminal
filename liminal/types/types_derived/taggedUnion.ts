import type { JSONKey } from "../../util/JSONKey.ts"
import type { JSONConstType } from "../const.ts"
import { type JSONObjectType, object } from "../object.ts"
import type { JSONStringType } from "../string.ts"
import { type Type } from "../Type.ts"
import { type NormalizeTypeLikeJ, type NormalizeTypeLikeT, type TypeLike } from "../TypeLike.ts"
import { _union, type JSONUnionType, union } from "../union.ts"

export function taggedUnion<const M extends TaggedUnionMembers, K extends Extract<keyof M, string>>(members: M): Type<
  {
    [L in K]: {
      type: L
      value: NormalizeTypeLikeT<M[L]>
    }
  }[K],
  JSONUnionType<
    {
      [L in K]: JSONObjectType<{
        type: JSONConstType<JSONStringType, L>
        value: NormalizeTypeLikeJ<M[L]>
      }>
    }[K]
  >
> {
  return union(
    ...Object.entries(members).map(([type, value]) => object({ type, value })),
  ) as never
}

export type TaggedUnionMembers = Record<JSONKey, TypeLike>
