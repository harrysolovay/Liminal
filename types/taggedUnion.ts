import { Ty } from "./Ty.ts"

export function taggedUnion<M extends Record<string, Ty | null>>(members: M): Ty<
  { [K in keyof M]: M[K] extends Ty ? { type: K; value: M[K]["T"] } : { type: K } }[keyof M],
  Exclude<M[keyof M], null>["P"],
  false
> {
  const entries = Object.entries(members)
  return Ty(
    (subschema) => ({
      discriminator: "type",
      anyOf: entries.map(([k, v]) => ({
        type: "object",
        properties: {
          type: {
            type: "string",
            const: k,
          },
          ...(v === null ? {} : {
            value: subschema(v),
          }),
        },
        required: ["type", ...v === null ? [] : ["value"]],
        additionalProperties: false,
      })),
    }),
    false,
    (v) => members[v.type]![""].transform(v),
  )
}
