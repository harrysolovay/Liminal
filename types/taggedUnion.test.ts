import { taggedUnion } from "./taggedUnion.ts"
import { assertTySnapshot } from "test_util"
import { boolean, number, string } from "./leaves.ts"

Deno.test("tagged union", async (t) => {
  await assertTySnapshot(
    t,
    taggedUnion({
      A: boolean,
      B: number,
      C: string,
    }),
  )
})
