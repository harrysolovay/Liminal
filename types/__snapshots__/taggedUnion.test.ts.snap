export const snapshot = {};

snapshot[`tagged union 1`] = `
{
  additionalProperties: false,
  description: "",
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              description: "",
              type: "boolean",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              description: "",
              type: "number",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              description: "",
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`tagged union 2`] = `
{
  additionalProperties: false,
  description: "",
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              description: "",
              type: "boolean",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              description: "",
              type: "number",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              description: "",
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "Description one.",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;

snapshot[`tagged union 3`] = `
{
  additionalProperties: false,
  description: "",
  properties: {
    value: {
      anyOf: [
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "A",
              type: "string",
            },
            value: {
              description: "",
              type: "boolean",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "B",
              type: "string",
            },
            value: {
              description: "",
              type: "number",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
        {
          additionalProperties: false,
          properties: {
            type: {
              const: "C",
              type: "string",
            },
            value: {
              description: "",
              type: "string",
            },
          },
          required: [
            "type",
            "value",
          ],
          type: "object",
        },
      ],
      description: "Description two. Description one.",
      discriminator: "type",
    },
  },
  required: [
    "value",
  ],
  type: "object",
}
`;
